### PR TITLE
feat: Vertex Vizier support in SDK.

### PR DESCRIPTION
**Pull request by halio-g**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue] b/230767019
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes # b/230767019 🦕

### CODE REVIEW

File: google/cloud/aiplatform/vizier/client_abc.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+to be `StudyInterface` or `TrialInterface` when you want to guarantee that
+a code block is cross-platform.
+
+Keywords:
+
+#Materialize: The method returns a deep copy of the underlying pyvizier object.
+Modifying the returned object does not update the Vizier service.
+"""
+
+# TODO(b/182496749): Add a dedicated .md file with more code examples.
```
Comment (sagipe) [inline google/cloud/aiplatform/vizier/client_abc.py:16]: Should this b/ be scrubbed out?

File: google/cloud/aiplatform/vizier/study.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            sc = pyvizier.StudyConfig()
+            sc.algorithm = pyvizier.Algorithm.RANDOM_SEARCH
+            sc.metric_information.append(
+                pyvizier.MetricInformation(
+                    name='pr-auc', goal=pyvizier.ObjectiveMetricGoal.MAXIMIZE))
+            root = sc.search_space.select_root()
+            root.add_float_param(
+                'learning_rate', 0.00001, 1.0, scale_type=pyvizier.ScaleType.LINEAR)
+            root.add_categorical_param('optimizer', ['adagrad', 'adam', 'experimental'])
+            study = aiplatform.Study.create_or_load(display_name='display_name', problem=sc)
```
Comment (sagipe) [inline google/cloud/aiplatform/vizier/study.py:104]: Perhaps: display_name='tuning study'?

File: google/cloud/aiplatform/vizier/study.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        try:
+            study = api_client.create_study(
+                parent=initializer.global_config.common_location_path(
+                    initializer.global_config.project,
+                    initializer.global_config.location,
+                ),
+                study=study,
+            )
+        except exceptions.AlreadyExists:
+            _LOGGER.info("The study is aleady created. Using existing study.")
```
Comment (sagipe) [inline google/cloud/aiplatform/vizier/study.py:129]: aleady -> already

File: google/cloud/aiplatform/vizier/study.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        list_optimal_trials_request = {"parent": self.resource_name}
+        optimal_trials_response = self.api_client.list_optimal_trials(
+            request=list_optimal_trials_request
+        )
+        return [Trial(trial.name) for trial in optimal_trials_response.optimal_trials]
+
+    def materialize_study_config(self) -> vz.StudyConfig:
+        """#Materializes the study config."""
+        study = self.api_client.get_study(name=self.resource_name)
+        return copy.deepcopy(vz.StudyConfig.from_proto(study.study_spec))
```
Comment (sagipe) [inline google/cloud/aiplatform/vizier/study.py:174]: Do we need the copy.deepcopy?
We're creating a new object here on every call

File: google/cloud/aiplatform/vizier/study.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    def from_uid(cls: Type[_T], uid: str) -> _T:
+        """Fetches an existing study from the Vizier service.
+
+        Args:
+          uid: Unique identifier of the study.
+        """
+        return Study(study_id=uid)
+
+    def suggest(
+        self, *, count: Optional[int] = None, worker: str = ""
```
Comment (sagipe) [inline google/cloud/aiplatform/vizier/study.py:186]: This argument has been renamed to client_id in the most recent version of client_abc.py (check Piper).
Do you want to override with the most recent version to make sure the abstractions match?

File: google/cloud/aiplatform/vizier/trial.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+        * If `measurement` is provided, then Vizier writes it as the trial's final
+        measurement and returns it.
+        * If `infeasible_reason` is provided, `measurement` is not needed.
+        * If neither is provided, then Vizier selects an existing (intermediate)
+        measurement to be the final measurement and returns it.
+
+        Args:
+          measurement: Final measurement.
+          infeasible_reason: Infeasible reason for missing final measurement.
```
Comment (sagipe) [inline google/cloud/aiplatform/vizier/trial.py:132]: Indefeasibly reason for...

File: google/cloud/aiplatform/vizier/study.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    _getter_method = "get_study"
+    _list_method = "list_studies"
+    _delete_method = "delete_study"
+    _parse_resource_name_method = "parse_study_path"
+    _format_resource_name_method = "study_path"
+
+    def __init__(
+        self,
+        study_id: str,
+        project: Optional[str] = None,
```
Comment (chansoo-google) [inline google/cloud/aiplatform/vizier/study.py:52]: Do you want to make the arguments keyword only by adding an asterik? https://screenshot.googleplex.com/7GGggHAfRd4J6oS

File: google/cloud/aiplatform/vizier/study.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        )
+        _LOGGER.log_action_started_against_resource_with_lro(
+            "Suggest", "study", self.__class__, suggest_trials_lro
+        )
+        _LOGGER.info(self.client_class.get_gapic_client_class())
+        trials = suggest_trials_lro.result()
+        _LOGGER.log_action_completed_against_resource("study", "suggested", self)
+        return [Trial(trial.name) for trial in trials.trials]
+
+    def delete(self) -> None:
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/vizier/study.py:211]: There's a shared `delete` implementation in the base class.

File: google/cloud/aiplatform/vizier/study.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        trials_response = self.api_client.list_trials(request=list_trials_request)
+        return [Trial(trial.name) for trial in trials_response.trials]
+
+    def optimal_trials(self) -> Collection[TrialInterface]:
+        """Returns optimal trial(s)."""
+        list_optimal_trials_request = {"parent": self.resource_name}
+        optimal_trials_response = self.api_client.list_optimal_trials(
+            request=list_optimal_trials_request
+        )
+        return [Trial(trial.name) for trial in optimal_trials_response.optimal_trials]
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/vizier/study.py:169]: There is a `_construct_sdk_resource_from_gapic` classmethod that can be used to avoid the extra GET call in the constructor of `Trial`.

File: google/cloud/aiplatform/vizier/study.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+            )
+        )
+
+    def trials(
+        self, trial_filter: Optional[vz.TrialFilter] = None
+    ) -> Collection[TrialInterface]:
+        """Fetches a collection of trials."""
+        list_trials_request = {"parent": self.resource_name}
+        trials_response = self.api_client.list_trials(request=list_trials_request)
+        return [Trial(trial.name) for trial in trials_response.trials]
```
[the remark this answers is not in the document]
Comment (sasha-gitg) [reply google/cloud/aiplatform/vizier/study.py:161]: Credentials should be forwarded in these scenarios.

File: google/cloud/aiplatform/vizier/trial.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        trial = self.api_client.get_trial(name=self.resource_name)
+        return vz.TrialConverter.from_proto(trial).parameters
+
+    @property
+    def status(self) -> vz.TrialStatus:
+        """Status of the Trial."""
+        trial = self.api_client.get_trial(name=self.resource_name)
+        return vz.TrialConverter.from_proto(trial).status
+
+    def delete(self) -> None:
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/vizier/trial.py:112]: Can also use delete implementation in the base class here.

File: tests/system/aiplatform/test_vizier.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+from google.api_core import exceptions
+from google.cloud import aiplatform
+from tests.system.aiplatform import e2e_base
+from google.cloud.aiplatform.vizier import pyvizier
+
+_TEST_STUDY_ID = 123
+
+
+class TestVizier(e2e_base.TestEndToEnd):
```
Comment (sasha-gitg) [inline tests/system/aiplatform/test_vizier.py:11]: Decorate with `tear_down_resources` to clean up in case the test fails. example: https://github.com/googleapis/python-aiplatform/blob/main/tests/system/aiplatform/test_experiments.py#L41

**Review by sasha-gitg — DISMISSED** [review]

File: google/cloud/aiplatform/vizier/client_abc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (halio-g) [reply google/cloud/aiplatform/vizier/client_abc.py:16]: Removed.

File: google/cloud/aiplatform/vizier/study.py (hunk 3 above)
Comment (halio-g) [reply google/cloud/aiplatform/vizier/study.py:129]: Done.

File: google/cloud/aiplatform/vizier/study.py (hunk 4 above)
Comment (halio-g) [reply google/cloud/aiplatform/vizier/study.py:174]: I think that's intended. Chansoo mentioned that there's a case that the existing study_config might be used to create another studies. If the customer doesn't want to copy the new object, they can use the get_study to get  the study_spec.

**Review by sagipe — APPROVED** [review]

File: google/cloud/aiplatform/vizier/pyvizier/automated_stopping.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+from google.cloud.aiplatform.compat.types import study as study_pb2
+
+AutomatedStoppingConfigProto = Union[
+    study_pb2.StudySpec.DecayCurveAutomatedStoppingSpec,
+    study_pb2.StudySpec.MedianAutomatedStoppingSpec,
+]
+
+
+@attr.s(frozen=True, init=True, slots=True, kw_only=True)
```
Comment (chansoo-google) [inline google/cloud/aiplatform/vizier/pyvizier/automated_stopping.py:15]: Please run a global replace 
attr.s -> attr.define 
attr.ib -> attr.field

So we can take advantage of the newest features.

**Review by rosiezou — DISMISSED** [review]

**Review by rosiezou — APPROVED** [review]

File: setup.py [hunk 14]
```diff
@@ -58,6 +58,19 @@
 datasets_extra_require = [
     "pyarrow >= 3.0.0, < 8.0dev",
 ]
+
+vizier_extra_require = [
+    "attrs==21.4.0",
+    "absl-py>=0.7",
+    "numpy>=1.19.0",
+    "protobuf>=3.6,<4.0",
+    "keras-tuner>=1.0,<2.0",
```
Comment (Ark-kun) [inline setup.py:67]: Where is keras-tuner used?

File: setup.py [hunk 15]
```diff
@@ -58,6 +58,19 @@
...
+    "portpicker==1.3.1",
+    "googleapis-common-protos==1.56.0",
```
Comment (Ark-kun) [inline setup.py:69]: Should these versions be so strict?

File: setup.py [hunk 16]
```diff
@@ -58,6 +58,19 @@
...
+    "google-api-python-client==1.12.8",
+    "google-vizier==0.0.3a",
```
Comment (Ark-kun) [inline setup.py:71]: Are there any issues with supporting higher versions of vizier? 0.0.3a seems to be quite behind 0.0.15
